server.js

```diff
@@ -1,38 +1,258 @@
+  const pageNumber = +page || 1;
+  const pageSize = 20;
+
+  // skip: E.g. page 3: 20 * (3-1) = 40, sends 40 as parameter to .skip()
+  // skips index 0-39 so that page 3 starts with the book that has index 40
+  const skip = pageSize * (pageNumber - 1);
+
+  // Books with filters based on title and/or lang query
+  // Paginated using limit and skip, change page using page query
+  const books = await Book.find({
+    title: new RegExp(title, 'i'),
+    language_code: new RegExp(lang, 'i')
+  })
+    .populate('authors')
+    .sort({ average_rating: -1 }) // Hard coded to sort data descending by average_rating.
+    .limit(pageSize)
+    .skip(skip);
+
+  // Used to display total books, without pagination
+  const booksNoPagination = await Book.find({
+    title: new RegExp(title, 'i'),
+    language_code: new RegExp(lang, 'i')
+  });
+  const totalNumberOfBooks = booksNoPagination.length;
+
+  const returnObject = {
+    totalNumberOfBooks: totalNumberOfBooks,
+    totalNumberOfPages: Math.ceil(totalNumberOfBooks / pageSize), // Round a number upward to its nearest integer
+    currentPage: pageNumber,
+    pageSize: pageSize,
+    skip: skip,
+    results: books
+  };
+
+  // If the result is zero, status set to 404 and returning a useful data in the response
+  if (books.length === 0) {
+    res.status(404).send({
+      error: 'Sorry, no books where found, please try a different query.'
+    });
+  } else {
+    res.json(returnObject);
+  }
+});
+
+// /books/top-20-rated endpoint
+// RETURNS: A list of 20 top rated books from mongoDB
+app.get('/books/top-20-rated', async (req, res) => {
+  const top20Rated = await Book.find()
+    .populate('authors')
+    .sort({ average_rating: -1 }) // Hard coded to sort data descending by average_rating.
+    .limit(20); // Hard coded to 20
+
+  if (top20Rated.length === 0) {
+    res.status(404).send({
+      error: 'Sorry, no books where found.'
+    });
+  } else
+    res.json({
+      topTwentyRated:
+        'A list of 20 top rated books sorted in descending by average_rating',
+      totalNumberOfBooks: top20Rated.length,
+      results: top20Rated
+    });
+});
+
+// /books/:id endpoint
+// RETURNS: A single book by id from mongoDB
+app.get('/books/:id', async (req, res) => {
+  const { id } = req.params;
+
+  try {
+    const singleBook = await Book.findOne({ bookID: id });
+
+    if (singleBook) {
+      res.json(singleBook);
+    } else res.status(404).send({ error: `No book with id: ${id} found.` });
+  } catch {
+    res.status(400).send({ error: `${id} is not a valid bookID.` });
+  }
+});
+
+// /authors endpoint
+// RETURNS: A list of authors from mongoDB
+//
+// PARAMETERS:
+//  - author
+//     usage: //authors/?author=douglas
+app.get('/authors', async (req, res) => {
+  const { author } = req.query;
+  const authorsArray = await Author.find({
+    authors: new RegExp(author, 'i')
+  });
+
+  if (!authorsArray) {
+    res.status(404).send({ Error: 'something went wrong' });
```

Comment:
Perhaps you could clarify this error-message a bit? Since the statuscode is a 404, maybe you could write "No author found" or something?